python/paddle/fluid/reader.py

```diff
@@ -348,6 +354,18 @@ def __init__(self,
         assert timeout >= 0, "timeout should be a non-negative value"
         self.timeout = timeout
 
+        if isinstance(dataset, IterableDataset):
+            self.dataset_kind = _DatasetKind.ITER
+            if shuffle:
+                raise ValueError(
+                    "IterableDataset not support shuffle, but got shuffle={}".
+                    format(shuffle))
+            if batch_sampler is not None:
+                raise ValueError(
+                    "IterableDataset expect unspecified batch_sampler")
+        else:
+            self.dataset_kind = _DatasetKind.MAP
+
         if batch_sampler is not None:
             assert isinstance(batch_sampler, BatchSampler), \
                 "batch_sampler should be None or subclass instance " \
```

Comment:
Can we remove the BatchSampler later, maybe it can also be `Iterable` object.

Additionally, can we support  specified `batch_sampler`  for IterableDataset later.  It seems that users can't custom sampling or batch strategies even by  themselves, since we can only support Iterable data with `IterableDataset` and `_InfiniteIterableSampler`

And would we also support Sampler except BatchSampler later

Reply:
BatchSampler can be custom in map-style dataset(implement `__getitem__`), for `IterableDataset`， which can only get sample sequencely， I couldn't think of scenarios that require `batch_sampler` customization, sure it should be support if there is customization requirements.

Sampler is mostly a sub-function of BatchSampler, IMHO, custom Sampler can be defined in custom BatchSampler?

Reply:
An example for batch_sampler customization is Transformer, it  changes the batch size counter by using word number rather than sentence number ,  currently it uses a map-style dataset . 

When custom batching strategies is needed, then Sampler may be abstracted from BatchSampler to reuse the sampling strategies.

However, it doesn't bother and we can consider it later.

Reply:
Thanks, I'll try to do some research and try to add this later~